gflownet/envs/cube.py

```diff
@@ -0,0 +1,1317 @@
+"""
+Classes to represent hyper-cube environments
+"""
+import itertools
+from abc import ABC, abstractmethod
+from typing import List, Optional, Tuple
+
+import matplotlib.pyplot as plt
+import numpy as np
+import numpy.typing as npt
+import pandas as pd
+import torch
+from sklearn.neighbors import KernelDensity
+from torch.distributions import Bernoulli, Beta, Categorical, MixtureSameFamily, Uniform
+from torchtyping import TensorType
+
+from gflownet.envs.base import GFlowNetEnv
+from gflownet.utils.common import copy, tbool, tfloat
+
+
+class Cube(GFlowNetEnv, ABC):
```

Comment:
Initially there was a version that only increments one dimension at a time. It would be good to have eventually. I have removed it for now. I'm good with the suggestion of `AbstractCube`.